### ISSUE TITLE
Scala.js 0.6.20 breaks Java 7 builds (NoSuchMethodError: java.util.concurrent.ConcurrentHashMap.keySet)

### ISSUE THREAD

**Issue by xuwei-k** [scala-js/scala-js#3128]
- https://github.com/xuwei-k/scalaz/commit/5f5bc477c7852ee5ac3ad5842eb0048135813f2c
- https://github.com/scala-js/scala-js/blob/v0.6.20/test-common/src/main/scala/org/scalajs/testcommon/RPCCore.scala#L155
- https://github.com/scala-js/scala-js/pull/3040
- https://gist.github.com/AlainODea/1375759b8720a3f9f094
- https://travis-ci.org/xuwei-k/scalaz/jobs/271913650#L1455

```
java.lang.NoSuchMethodError: java.util.concurrent.ConcurrentHashMap.keySet()Ljava/util/concurrent/ConcurrentHashMap$KeySetView;
	at org.scalajs.testcommon.RPCCore.close(RPCCore.scala:155)
	at org.scalajs.testadapter.ComJSEnvRPC.close(ComJSEnvRPC.scala:59)
	at org.scalajs.testadapter.ScalaJSRunner$$anonfun$waitAndClose$1$1.apply(ScalaJSRunner.scala:69)
	at org.scalajs.testadapter.ScalaJSRunner$$anonfun$waitAndClose$1$1.apply(ScalaJSRunner.scala:69)
	at scala.util.Success$$anonfun$map$1.apply(Try.scala:206)
	at scala.util.Try$.apply(Try.scala:161)
	at scala.util.Success.map(Try.scala:206)
	at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:235)
	at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:235)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
	at scala.concurrent.impl.ExecutionContextImpl$$anon$3.exec(ExecutionContextImpl.scala:107)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
java.lang.NoSuchMethodError: java.util.concurrent.ConcurrentHashMap.keySet()Ljava/util/concurrent/ConcurrentHashMap$KeySetView;
	at org.scalajs.testcommon.RPCCore.close(RPCCore.scala:155)
	at org.scalajs.testadapter.ComJSEnvRPC.close(ComJSEnvRPC.scala:59)
	at org.scalajs.testadapter.ScalaJSRunner$$anonfun$waitAndClose$1$1.apply(ScalaJSRunner.scala:69)
	at org.scalajs.testadapter.ScalaJSRunner$$anonfun$waitAndClose$1$1.apply(ScalaJSRunner.scala:69)
	at scala.util.Success$$anonfun$map$1.apply(Try.scala:206)
	at scala.util.Try$.apply(Try.scala:161)
	at scala.util.Success.map(Try.scala:206)
	at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:235)
	at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:235)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
	at scala.concurrent.impl.ExecutionContextImpl$$anon$3.exec(ExecutionContextImpl.scala:107)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by sjrd** [comment]
Huh! The result type of `keySet()` was refined from [`Set<K>` in JDK 7](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ConcurrentHashMap.html#keySet()) down to [`ConcurrentHashMap.KeySetView<K,V>` in JDK 8](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#keySet--). Since we build our distribution on JDK 8, the compiler resolves to the more specific type instead of the binary compatible synthetic bridge.

I'm not sure how to fix that without avoiding `keySet()` altogether.

**Comment by sjrd** [comment]
The gist linked above does mention a sidestepping workaround which we could use. It is probably the most reasonable solution in our case. Setting up the correct bootclasspath during compilation is going to be very fragile, as it is virtually impossible to write down in the build in a platform-independent way, so we risk manually doing it wrong (or just forget) at the time we build.

**Comment by sjrd** [comment]
Fixed in cb194ae118c02744c52bc9be268602425fbf56e8.
